functions/snippets/src/main/java/GcsEvent.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START functions_helloworld_gcs_event]
+public class GcsEvent {
```

Comment:
This class has to be deserializable using Gson, given its use in `BackgroundFunction<GcsEvent>`. Though private fields can be made to work with appropriate Gson annotations, I'm not sure that will actually work at the moment for obscure ClassLoader reasons, and it would certainly be much more complicated than exposing mutable fields like this. However, it's a bit strange to me that the class is public but its fields are not. I think I'd make them both package-private (i.e. drop the `public` keyword here).

Reply:
@eamonnmcmanus Any particular reason we are having the user define their own class rather than provide one as part of the functions framework? I would imagine it's largely static? Seems a bit boilerplate and introduces a chance for error.

Reply:
These fields should be private and the class should use `getters` to access them.

There should also be a comment explaining that the functions environment will use GSON to serialize them, and that it is important the field names are exactly as provided (or else it won't work). 

Additionally, I think you should apply Eamon's comment about making the class being package-private.

Reply:
As I mentioned in my earlier comment, I'm not sure private fields will work. They might also be a little misleading, in that the fields are in fact set (by Gson) from outside the class. We do generally avoid mutable non-private fields, but in this case of a pure-data class with no methods they're relatively inoffensive.

I can do some more research and document the alternatives here. The situation is further complicated by our future direction of using CloudEvents. The current CloudEvents Java SDK uses Jackson rather than Gson, and by default requires the fields to be public.

Reply:
Apologies - rereading my previous comment I realized I had misread part of your comment. 

From my experience, Gson doesn't require any special annotations for private fields. According to the [user guide](https://sites.google.com/site/gson/gson-user-guide):
```
It is perfectly fine (and recommended) to use private fields
There is no need to use any annotations to indicate a field is to be included for serialization and deserialization. All fields in the current class (and from all super classes) are included by default.
```

I agree that adding a constructor could be misleading, which is why I would like to make sure each example of these classes has a clear comment explaining how it is being used. An alternative to using a constructor would be to use Gson to marshal a json string into the object instead (which would probably help crystalize how the event is being treated). I suggested the constructor because I thought it was simpler. 

I'm less familiar with Jackson, but it seems hard to believe that it doesn't support immutable POJOs. (I did some searching - it looks like it might use a provided constructor. Seems like another reason to add it.)

Finally, I do realize that it's unlikely that we are going to get in to trouble with package-private fields in a small sample like this. However, it's a best practice to always use getters and setters, even with POJOs. These samples are going to be read copied from developers with variable levels of experience from the freshman to the senior software engineer. It's important that we make sure to re-enforced idiomatic practices so we don't accidentally teach the wrong thing to someone who doesn't know any better.